### ISSUE TITLE
$http service: specification of `status` in error case

### ISSUE THREAD

**Issue by jgehrcke** [angular/angular.js#11945]
https://docs.angularjs.org/api/ng/service/$http clearly states that the `status` argument to the `success` handler is between 200 and 299. In all other cases, the `error` handler is called. However, there is no statement about the values that `status` can assume within the error handler. I guess that it is either `0` (if no back-end response arrived, for whatever reason) or the HTTP status indicated by the back-end. I hope no other values are allowed (such as `null`). Is that correct? Is that guaranteed? If it is, it should be documented, correct?

Thanks,

Jan-Philip

**Comment by gkalpak** [comment]
I think, the only guarantee is that (if it's a number) it will be <= 0. Other than that, it comes directly from the `XMLHttpRequest`'s `status` property, which [according to MDN][1] should be an _unsigned short_, but I am not sure what some weird backend might do (and what browsers would accept).

It's extremely unlikely that you get a non-numeric value (but it's not something Angular-specific really).

[1]: https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest#xmlhttprequest-status

**Comment by jgehrcke** [comment]
You meant write `>= 0`, right?

So, if AngularJS guarantees to not touch and just forward `XMLHttpRequest.status`, then things are clear enough. The behavior of this flag is specified [here](http://www.w3.org/TR/XMLHttpRequest/#the-status-attribute
): 
```text
The status attribute must return the result of running these steps:

    If the state is UNSENT or OPENED, return 0.

    If the error flag is set, return 0.

    Return the HTTP status code.
```

In other words, one can safely expect this to be an integer, and checking for 0 is a useful way to catch all errors that were not produced by the back-end itself.

**Comment by gkalpak** [comment]
@jgehrcke, yeah I meant `>= 0`. 

<sub>BTW, I kind of lied, Angular does a tiny little manipulation of `XMLHttpRequest.status`: It passes it through `Math.max(status, 0)` (in order to _"normalize internal statuses"_).</sub>
<sub>This has the side-effect that `status` is parsed as number, so things might get weird if the backend returns a non-numeric status value. (But if you have a non-number-status-returning backend, things are weird already, so...)</sub>

--
**Bottom line:**
It is (reasonably) safe to assume, that the `status` passed to the error handler will be a non-negative integer outside of the [200, 300) range.

**Comment by zachcarroll** [comment]
@gkalpak @jgehrcke 
I'm seeing an example where the status, as displayed in the network tab, shows a 404, but the code coming through on the response.status field is `-1`.  

Does anyone have an explanation of this? And can anyone tell me how to get access to the real status code coming through on the response (in this case, the 404).

**Comment by gkalpak** [comment]
There's been a change recently, so now -1 is also preserved. (I think it applies to requests that timed out or where cancelled.)

I doubt it is possible to get a status of -1 for a request that resulted in a 404 (but maybe it's a bug).
Browsers do sometimes report weird things. @zachcarroll, could you provide a live demo of the issue ?

**Comment by zachcarroll** [comment]
@gkalpak thanks for the response.  I can't provide specific example of what I'm seeing cause its proprietary code.   But, I just attempted to reproduce using public spotify api in a codepen and its working just fine.

So, this leads me to believe that its something server-side that's not setup properly.  I'm hitting a real server url, at an endpoint location that doesn't exist (for example: foobar.com/api/non-existent-url).  In the console it looks to be returning a 404, but maybe this is just the browser interpreting?  Does this response code need to be explicitly setup on the server-side?

**Comment by gkalpak** [comment]
I am not aware of all the weird ways browsers choose to interpret or display things (e.g. I've seen misconfigured CORS requests reporting a status of -1, while the browser shows 200).

If I was trying to debug your issue, I would try with raw `XMLHttpRequest` and see what the response(/status/statusText) looks like (and compare it to what the browser shows in the Network tab).
If the results are the same as with `$http`, then it's not Angular-specific and you can debug it from there (having removed one dependency layer - yay !).

**Comment by zachcarroll** [comment]
thanks a lot, ill try that now

**Comment by zachcarroll** [comment]
@gkalpak so i updated the call to use jQuery $.ajax instead, and checked the status coming through there. its set to 0.  So is it possible angular interprets status code of 0 as -1?  Either way, looks like server-side error codes haven't been configured properly yet...

**Comment by gkalpak** [comment]
@zachcarroll, Angular (at least the latest versions) is not converting 0 status to -1. It just converts <-1 to 0 or leaves it unchanged.
It might be the other way around (jQuery "normalizing" the status) - I don't know.

BTW, `$.ajax` is **not** what I meant by _raw XMLHttpRequest_ :)

**Comment by AbhishekGarg** [comment]
PR #14459

**Comment by AbhishekGarg** [comment]
@gkalpak I guess we can close this issue safely :relaxed:
